File: android/app/src/main/java/com/tandem/emt/features/incidentReport/IncidentReportRepository.kt
package com.tandem.emt.features.incidentReport

import com.tandem.emt.features.incidentReport.models.IncidentReport
import com.tandem.emt.features.incidentReport.models.ReportStatus
import kotlinx.coroutines.delay

interface IncidentReportRepository {
    suspend fun submitReport(report: IncidentReport): Result<IncidentReport>
}

class IncidentReportRepositoryImpl : IncidentReportRepository {
    override suspend fun submitReport(report: IncidentReport): Result<IncidentReport> {
        // Stub: simulate network delay and return the report as submitted
        delay(1000)
        return Result.success(
            report.copy(
                status = ReportStatus.SUBMITTED,
                updatedAt = System.currentTimeMillis()
            )
        )
    }
}

class FakeIncidentReportRepository : IncidentReportRepository {
    var submitResult: Result<IncidentReport> = Result.success(
        IncidentReport(
            id = "mock-id",
            patientName = "",
            patientAge = null,
            chiefComplaint = "",
            vitalSigns = com.tandem.emt.features.incidentReport.models.VitalSigns(),
            treatments = emptyList(),
            procedures = emptyList(),
            disposition = null,
            transportDestination = null,
            narrative = "",
            timestamps = com.tandem.emt.features.incidentReport.models.ReportTimestamps(),
            status = ReportStatus.SUBMITTED,
            createdAt = System.currentTimeMillis(),
            updatedAt = System.currentTimeMillis()
        )
    )

    override suspend fun submitReport(report: IncidentReport): Result<IncidentReport> {
        return submitResult
    }
}

File: ios/TandemEMT/Features/IncidentReport/IncidentReportRepository.swift
import Foundation

protocol IncidentReportRepositoryProtocol: Sendable {
    func submitReport(_ report: IncidentReport) async throws -> IncidentReport
}

final class IncidentReportRepository: IncidentReportRepositoryProtocol {
    private let apiClient: APIClientProtocol

    init(apiClient: APIClientProtocol = APIClient.shared) {
        self.apiClient = apiClient
    }

    func submitReport(_ report: IncidentReport) async throws -> IncidentReport {
        // Stub: simulate network delay and return the report as submitted
        try await Task.sleep(nanoseconds: 1_000_000_000)
        var submitted = report
        submitted.status = .submitted
        submitted.updatedAt = Date()
        return submitted
    }
}

final class MockIncidentReportRepository: IncidentReportRepositoryProtocol, @unchecked Sendable {
    var submitResult: Result<IncidentReport, Error> = .success(
        IncidentReport(
            id: "mock-id",
            patientName: "",
            patientAge: nil,
            chiefComplaint: "",
            vitalSigns: VitalSigns(),
            treatments: [],
            procedures: [],
            disposition: nil,
            transportDestination: nil,
            narrative: "",
            timestamps: ReportTimestamps(),
            status: .submitted,
            createdAt: Date(),
            updatedAt: Date()
        )
    )

    func submitReport(_ report: IncidentReport) async throws -> IncidentReport {
        return try submitResult.get()
    }
}
